### PR TITLE
Operator arguments

### LINKED ISSUE #148: Operator arguments.
Currently we're parsing arguments that are expressions, but not ones that start with operators. This includes:

- [ ] Splat arguments that start with `*`.
- [ ] Double splat arguments that start with `**`.
- [ ] Block arguments that start with `&`.

This will involve updating the `parse_arguments` function.

### PR DESCRIPTION
**Pull request by DmitryTsepelev**
Closes #148. Please let me know if there are some scenarios missing (or we can add them later)

### CODE REVIEW

File: config.yml [hunk 1]
```diff
@@ -658,6 +658,42 @@ nodes:
             bar(...)
             ^^^^^^^^
           end
+  - name: SplatArgumentsNode
```
Comment (kddnewton) [inline config.yml:661]: We already have a `StarNode`, can you use that?

File: config.yml [hunk 2]
```diff
@@ -658,6 +658,42 @@ nodes:
...
+        type: token
+      - name: expression
+        type: node
+    location: operator->expression
+    comment: |
+      Represents splat method arguments.
+
+          bar(*args)
+          ^^^^^^^^^^
+  - name: DoubleSplatArgumentsNode
```
Comment (kddnewton) [inline config.yml:673]: Looks like folks are calling this `kwsplat` (parser gem/ripper). Could we call this `KeywordStarNode` to keep that spirit here?

File: config.yml [hunk 3]
```diff
@@ -658,6 +658,42 @@ nodes:
...
+        type: token
+      - name: expression
+        type: node
+    location: operator->expression
+    comment: |
+      Represents double splat method arguments.
+
+          bar(**args)
+          ^^^^^^^^^^^
+  - name: BlockArgumentsNode
```
Comment (kddnewton) [inline config.yml:685]: I think `BlockArgumentNode` (no s).

File: src/yarp.c [hunk 4]
```diff
@@ -2068,7 +2071,33 @@ parse_arguments(yp_parser_t *parser, yp_node_t *arguments, yp_token_type_t termi
...
+    }
+
+    yp_node_t *argument;
+
+    if (accept(parser, YP_TOKEN_STAR)) {
+      if (parsed_double_splat_argument) {
+        yp_error_list_append(&parser->error_list, "Unexpected splat argument after double splat.", parser->current.start - parser->start);
+      }
+      yp_token_t operator = parser->previous;
+      yp_node_t *value = parse_expression(parser, BINDING_POWER_NONE, "Expected to be able to parse an argument.");
```
Comment (kddnewton) [inline src/yarp.c:2085]: I don't think `BINDING_POWER_NONE` is correct here. That would accept

```ruby
foo(*bar and baz)
```

We should figure out the correct binding power. I've been doing this by manually binary searching through the list of potential values (defined in `yarp.c` `binding_powers` list).

File: src/yarp.c [hunk 5]
```diff
@@ -2068,7 +2071,33 @@ parse_arguments(yp_parser_t *parser, yp_node_t *arguments, yp_token_type_t termi
...
+      argument = yp_node_splat_arguments_node_create(parser, &operator, value);
+    } else if (accept(parser, YP_TOKEN_STAR_STAR)) {
+      yp_token_t operator = parser->previous;
+      yp_node_t *value = parse_expression(parser, BINDING_POWER_NONE, "Expected to be able to parse an argument.");
```
Comment (kddnewton) [inline src/yarp.c:2089]: Same here with the binding power.

File: src/yarp.c [hunk 6]
```diff
@@ -2068,7 +2071,33 @@ parse_arguments(yp_parser_t *parser, yp_node_t *arguments, yp_token_type_t termi
...
+      argument = yp_node_double_splat_arguments_node_create(parser, &operator, value);
+      parsed_double_splat_argument = true;
+    } else if (accept(parser, YP_TOKEN_AMPERSAND)) {
+      yp_token_t operator = parser->previous;
+      yp_node_t *value = parse_expression(parser, BINDING_POWER_NONE, "Expected to be able to parse an argument.");
```
Comment (kddnewton) [inline src/yarp.c:2094]: Same here with the binding power.

**Review by kddnewton — CHANGES_REQUESTED** [review]
A couple of changes, but looks good so far!

File: test/parse_test.rb [hunk 7]
```diff
@@ -1117,7 +1174,10 @@ class ParseTest < Test::Unit::TestCase
       IDENTIFIER("foo"),
       PARENTHESIS_LEFT("("),
       ArgumentsNode(
-        [StarNode(IDENTIFIER("rest"), CallNode(nil, nil, IDENTIFIER("rest"), nil, nil, nil, "rest"))]
+        [StarNode(
+          STAR("*"),
```
Comment (DmitryTsepelev) [inline test/parse_test.rb:1178]: not quite related, but operator should be a star, not `rest` here, right? cause we have `rest` in the call node

File: src/yarp.c [hunk 8]
```diff
@@ -2075,7 +2053,46 @@ parse_arguments(yp_parser_t *parser, yp_node_t *arguments, yp_token_type_t termi
...
+        if (!yp_token_list_includes(&parser->current_scope->as.scope.locals, &parser->previous)) {
+          yp_error_list_append(&parser->error_list, "unexpected * when parent method is not forwarding.", parser->previous.start - parser->start);
+        }
+        argument = yp_node_star_node_create(parser, &previous, NULL);
+      } else {
+        if (parsed_double_splat_argument) {
+          yp_error_list_append(&parser->error_list, "Unexpected splat argument after double splat.", parser->current.start - parser->start);
+        }
+        yp_token_t operator = parser->previous;
+        yp_node_t *value = parse_expression(parser, BINDING_POWER_DEFINED, "Expected to be able to parse an argument.");
```
Comment (DmitryTsepelev) [inline src/yarp.c:2074]: Using `BINDING_POWER_DEFINED` like it was in the `parse_argument` which was merged here to make working with flags easier.

**Review by DmitryTsepelev — COMMENTED** [review]
Updated everything, please check again

File: test/parse_test.rb (hunk 7 above)
Comment (kddnewton) [reply test/parse_test.rb:1178]: Yeah this is definitely incorrect as it currently stands

Comment (DmitryTsepelev) [reply test/parse_test.rb]: @kddnewton cool, resolved conflicts, ready for a next round of review 🙂

**Review by kddnewton — APPROVED** [review]
Nice work @DmitryTsepelev!

**Pull request merged**